### PR TITLE
Pusher on App Engine Standard tutorial

### PR DESCRIPTION
**Pull request by jabubake**
Currently focused on Java, Python addition : TBD

### CODE REVIEW

File: tutorials/appengine-pusher/index.md [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+
+    String body = CharStreams.readLines(request.getReader()).toString();
+    String json = body.replaceFirst("^\\[", "").replaceFirst("\\]$", "");
+    Map<String, String> data = gson.fromJson(json, typeReference.getType());
+    String message = data.get("message");
+    String socketId = data.get("socket_id");
+    String channelId = data.get("channel_id");
+
+    User user = UserServiceFactory.getUserService().getCurrentUser();
+    // user email prefix as display name for current logged in user
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Capitalize: "User"

Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "currently authenticated"

File: tutorials/appengine-pusher/index.md [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+    messageData.put("message", taggedMessage);
+
+    // Send a message over the Pusher channel (maximum size of a message is 10KB)
+    Result result =
+        PusherService.getDefaultInstance()
+            .trigger(
+                channelId,
+                "new_message", // name of event
+                messageData,
+                socketId); // (optional) use client socket_id to exclude the sender from receiving the message
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Capitalize "optional" and "use"

File: tutorials/appengine-pusher/index.md [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+    // result.getStatus() == SUCCESS indicates successful transmission
+    messageData.put("status", result.getStatus().name());
+
+    response.getWriter().println(gson.toJson(messageData));
+  }
+}
+```
+
+## Using Pusher on your client
+
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Avoid nesting headings. Add a sentence here explaining what the reader will learn. Something better than this: "The following sections explain how to use Pusher on your client..."

File: tutorials/appengine-pusher/index.md [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+### Connecting to Pusher
+
+Client connections require the application key to be provided. If the client will be subscribing
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "require an application key."

File: tutorials/appengine-pusher/index.md [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+to private or presence channels, a server-side authentication endpoint must be  provided as well.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Delete extra space.

File: tutorials/appengine-pusher/index.md [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+The client will attempt to use `/pusher/auth` path for the endpoint, if one is not explicitly provided.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "attempts to use"

Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Delete comma.

File: tutorials/appengine-pusher/index.md [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+
+The following example illustrates how to instantiate a Pusher connection using a custom authentication
+ endpoint. For more on client connections, refer to [this resource](https://pusher.com/docs/client_api_guide/client_connect).
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "For more information about connections,"

Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change to "this page"

File: tutorials/appengine-pusher/index.md [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+        encrypted: true
+    });
+```
+
+### Subscribe to a channel
+
+A client can subscribe to multiple channels. A subscription to a private or presence channel auto-triggers the
+authentication endpoint. 
+
+The following example illustrates subscribing to a channel.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change to colon.

File: tutorials/appengine-pusher/index.md [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+
+[embedmd]:# (java/src/main/webapp/WEB-INF/view/chat.jsp /\/\/ subscribe to the chat room/ /\);/)
+```jsp
+// subscribe to the chat room presence channel, eg. "presence-my-room"
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Capitalize "subscribe"

File: tutorials/appengine-pusher/index.md [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+    var channel = pusher.subscribe(channel_name);
+```
+
+### Bind to Pusher events
+
+On subscription success/error, Pusher sends events that a client can easily attach
+to an event callback as shown in this example.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Delete "as shown in this example"

File: tutorials/appengine-pusher/index.md [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+
+In the case of presence channels, Pusher sends additional events when a user connects 
+or disconnects from the channel.
+
+The following code snippets show to bind event handlers to Pusher events.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change to "events:"

File: tutorials/appengine-pusher/index.md [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+            // track member removals from channel
+            $('#chat_widget_member_' + member.id).remove();
+            updateOnlineCount();
+        });
+
+```
+
+### Bind to channel events
+
+[Event binding](https://pusher.com/docs/client_api_guide/client_events#bind-events) may be done bound to a channel
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Please explain what event binding is and why I might want to use it.

File: tutorials/appengine-pusher/index.md [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+or to all channels subscribed by a client.
+
+The following example shows attaching an event callback to a user triggered event.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change "shows" to "demonstrates"

Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change to "user-triggered"

File: tutorials/appengine-pusher/index.md [hunk 14]
```diff
@@ -0,0 +1,449 @@
...
+In the case of the chat application, the event callback will be used to update the messages
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change to "is used to update"

File: tutorials/appengine-pusher/index.md [hunk 15]
```diff
@@ -0,0 +1,449 @@
...
+                whosonline_html += '<li class="chat_widget_member" id="chat_widget_member_'
+                    +
+                    member.id + '">' + member.info.displayName + '</li>';
+            });
+```
+
+
+### Trigger server-side endpoint to send messages
+
+Now use a server-side endpoint as described earlier to trigger an event on a chat message
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "Now you can use a server-side endpoint..."

File: tutorials/appengine-pusher/index.md [hunk 16]
```diff
@@ -0,0 +1,449 @@
...
+ that is submitted.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Delete "that is submitted"

File: tutorials/appengine-pusher/index.md [hunk 17]
```diff
@@ -0,0 +1,449 @@
...
+                    } else {
+                        alert("Error sending chat message : " + msg.status);
+                    }
+                }, "json");
+
+            return false;
+        });
+```
+
+## Disconnecting from Pusher
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Hmm, looks like something went wrong with the formatting here, probably somewhere in the code snippet...

File: tutorials/appengine-pusher/index.md [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+
+Pusher automatically closes connections when a user navigates to another web page or closes their
+ web browser, this need n't be done manually.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Delete: ", this need n't be done manually"

File: tutorials/appengine-pusher/index.md [hunk 19]
```diff
@@ -0,0 +1,449 @@
...
+If required, can be done as described [here](https://pusher.com/docs/client_api_guide/client_connect#disconnecting).
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "If you need to close a client connection manually, refer to [this page](https://pusher.com/docs/client_api_guide/client_connect#disconnecting)."

File: tutorials/appengine-pusher/index.md [hunk 20]
```diff
@@ -0,0 +1,449 @@
...
+## Running the application locally
+  
+The application uses the [App Engine Maven Plugin](/appengine/docs/standard/java/tools/using-maven)
+to test and deploy to the Google Cloud App Engine Standard environment.
+
+  ```
+     mvn clean appengine:run
+  ```
+  
+Access [http://localhost:8080](http://localhost:8080) via the browser, login and join the chat room.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Add comma between "login" and "and join"

File: tutorials/appengine-pusher/index.md [hunk 21]
```diff
@@ -0,0 +1,449 @@
...
+
+The chat window will contain a link you can use to join the room as a different user in another browser.
+
+You should now be able to view both the users within the chat application window and send messages to one another.
+
+## Deploying
+  
+  - Deploy the application to the project
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Add colon: "project:"

File: tutorials/appengine-pusher/index.md (hunk 17 above)
Comment (jabubake) [inline tutorials/appengine-pusher/index.md]: This issue is still open, I'm looking into it.

File: tutorials/appengine-pusher/index.md [hunk 22]
```diff
@@ -0,0 +1,449 @@
...
+This allows for efficient event routing in the clients.
+Note: A subscriber will receive all messages published over a channel.
+
+Events may be trigged by the user or Pusher.
+In case of Pusher triggered events on a channel, the event name is
+prefixed with `pusher:`, for example: `pusher:subscription-succeeded`.
+
+## Chat application
+The sample application demonstrates presence channels in Pusher for a chat application:
+View complete source code [here](java).
```
Comment (jabubake) [inline tutorials/appengine-pusher/index.md]: Did you mean `(/java)` ? will that correctly point to the sub-folder java ? I could n't get it to work on testing.

File: tutorials/appengine-pusher/index.md (hunk 17 above)
Comment (jabubake) [inline tutorials/appengine-pusher/index.md]: @ryanmcginnis : I'm not seeing the issue you mention here : what exactly do you think is wrong ?  https://github.com/GoogleCloudPlatform/community/blob/appengine-pusher-tutorial/tutorials/appengine-pusher/index.md

Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: It isn't appearing for me now, but previously it appeared that all of the text after that code sample was in italics. It appears to be OK now though, so you should skip this issue.

File: tutorials/appengine-pusher/index.md [hunk 23]
```diff
@@ -0,0 +1,450 @@
...
+   clients. 
+    
+## Before you begin  
+
+1.  Create a project in the [Google Cloud Platform Console](https://console.cloud.google.com/).
+1.  Install the [Google Cloud SDK](https://cloud.google.com/sdk/) and run:
+    ```
+       gcloud init
+    ```
+    If this is your first time creating an App Engine application, run the following command
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: I suggest make this its own step. We have instructed the reader to create a new project, so it is the "first time" they will have created an App Engine application. If they meet these prerequisites, they will likely choose to skip this section.

If you agree, change sentence to:

"1. Create an App Engine application:"

File: tutorials/appengine-pusher/index.md [hunk 24]
```diff
@@ -0,0 +1,450 @@
...
+    to create a new application:
+    ```
+       gcloud app create
+    ```
+
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Was there a heading here before? All of the following sections are currently nested under "Before you begin" which doesn't seem right.

File: tutorials/appengine-pusher/index.md [hunk 25]
```diff
@@ -0,0 +1,450 @@
...
+### Setting up a Pusher account
+
+To set up a Pusher account, perform these steps:
+
+1. [Create a Pusher account](https://pusher.com/signup).
+1. Once signed in, you are directed to a [dashboard](https://dashboard.pusher.com/).
+   The dashboard provides a convenient way to retrieve application settings, view errors
+   and a console to debug calls to your application. From the dashboard's left panel,
+   create a new application using "Your apps" and note down the cluster,
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "create a new application using "Your apps" and note down the cluster, ..."

Change to: "Click __Your apps__ to create a new application. Copy the cluster, application ID, key, and secret for later use."

File: tutorials/appengine-pusher/index.md [hunk 26]
```diff
@@ -0,0 +1,450 @@
...
+```
+
+## Using Pusher on your client
+
+The following sections explain how to subscribe to Pusher channels on your client using the
+JavaScript Websocket SDK.
+
+### Connecting to Pusher
+
+Client connections require an application key to be provided. If the client will be subscribing
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Delete "to be provided"
Change "will be" to "is"

File: tutorials/appengine-pusher/index.md [hunk 27]
```diff
@@ -0,0 +1,450 @@
...
+to private or presence channels, a server-side authentication endpoint must be provided as well.
+The client attempts to use `/pusher/auth` path for the endpoint if one is not explicitly provided.
+
+The following example illustrates how to instantiate a Pusher connection using a custom authentication
+ endpoint. For more information about connections, refer to
+ [this page](https://pusher.com/docs/client_api_guide/client_connect).
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "For more information about connections, refer to [this page](https://pusher.com/docs/client_api_guide/client_connect)."

Change to:

"For more information about connections, refer to Pusher's [Connection](https://pusher.com/docs/client_api_guide/client_connect) documentation."

File: tutorials/appengine-pusher/index.md [hunk 28]
```diff
@@ -0,0 +1,450 @@
...
+### Send messages over a channel
+
+Authenticated clients can submit messages to the channel over HTTP using a server-side endpoint
+as shown in the following example.
+
+The sender can be [excluded](https://pusher.com/docs/server_api_guide/server_excluding_recipients)
+from receiving the broadcast message by passing in its own socket ID when triggering an event.
+Individual messages are limited to 10KB in size.
+
+For more on publishing messages to multiple channels or batching multiple messages, refer to
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "For more on publishing messages to multiple channels or batching multiple messages, refer to [this page](https://pusher.com/docs/server_api_guide/interact_rest_api#publishing-events)."

Change to:

"For more on publishing messages to multiple channels or batching multiple messages, refer to Pusher's [Publishing events](https://pusher.com/docs/server_api_guide/interact_rest_api#publishing-events) documentation."

File: tutorials/appengine-pusher/index.md [hunk 29]
```diff
@@ -0,0 +1,450 @@
...
+```
+
+### Authorize client subscriptions for secure channels
+
+Clients connecting to private or presence channels require server-side authentication.
+The authentication endpoint can be implemented in your server as shown in this example.
+The REST SDK provides methods to retrieve the required authentication JSON.
+
+Note: Private channels do not require user information to be provided as part of the authentication.
+To learn more about authentication in Pusher, see [Pusher's documentation](https://pusher.com/docs/authenticating_users).
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: "To learn more about authentication in Pusher, see [Pusher's documentation](https://pusher.com/docs/authenticating_users)"

Change to:

"To learn more about authentication in Pusher, refer to Pusher's [Authenticating users](https://pusher.com/docs/authenticating_users) documentation."

File: tutorials/appengine-pusher/index.md [hunk 30]
```diff
@@ -0,0 +1,450 @@
...
+   application id, key and secret.
+1. Update [appengine-web.xml](java/src/main/webapp/WEB-INF/appengine-web.xml) with your Pusher
+   account credentials.
+   
+### Pusher SDKs
+
+Pusher provides a [range of libraries in different languages](https://pusher.com/docs/libraries):
+
+- Their REST SDKs provide the ability to authenticate clients and 
+  publish HTTP events from your server.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change "Their" to "Pusher's"

File: tutorials/appengine-pusher/index.md [hunk 31]
```diff
@@ -0,0 +1,450 @@
...
+- The WebSocket SDKs handle subscription.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change "The" to "Pusher's"

File: tutorials/appengine-pusher/index.md (hunk 24 above)
Comment (jabubake) [inline tutorials/appengine-pusher/index.md]: Good catch, I had taken out and forgot to convert the headings following it to `##` from `###`. Please see if it looks better now.

**Comment by jabubake** [comment]
@ryanmcginnis  : fixed the issues you cited.

File: tutorials/appengine-pusher/index.md (hunk 24 above)
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Looks good!

File: tutorials/appengine-pusher/index.md [hunk 32]
```diff
@@ -0,0 +1,452 @@
...
+- **Private**: Server-side authentication is enforced, channel names must be prefixed with `private-`
+- **Presence**: Server-side authentication is enforced, channel names must be prefixed with `presence-`,
+   and all members can view who have connected/disconnected from the channel.
+
+## Events
+
+An [Event](https://pusher.com/docs/client_api_guide/client_events) is a  message with a named type.
+Custom event handlers can be attached to a given event type. 
+This allows for efficient event routing in the clients.
+Note: A subscriber will receive all messages published over a channel.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Bold "Note:"

File: tutorials/appengine-pusher/index.md [hunk 33]
```diff
@@ -0,0 +1,452 @@
...
+## Using Pusher on your server
+The server-side REST SDK is used to initialize a Pusher instance, 
+authorize secure presence channels, and provide clients an endpoint to trigger events.
+
+### Connecting to Pusher
+The following code provides examples of initializing and connecting to Pusher.
+Use the credentials from the application you created to initialize and connect to Pusher,
+as shown in this example.
+
+Note: It is important to provide the cluster information if not using the default `mt1` (`us-east-1`) cluster.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Bold "Note:"

File: tutorials/appengine-pusher/index.md [hunk 34]
```diff
@@ -0,0 +1,452 @@
...
+}
+```
+
+### Authorize client subscriptions for secure channels
+
+Clients connecting to private or presence channels require server-side authentication.
+The authentication endpoint can be implemented in your server as shown in this example.
+The REST SDK provides methods to retrieve the required authentication JSON.
+
+Note: Private channels do not require user information to be provided as part of the authentication.
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Bold "Note:"

File: tutorials/appengine-pusher/index.md [hunk 35]
```diff
@@ -0,0 +1,452 @@
...
+            return false;
+        });
+```
+
+## Disconnecting from Pusher
+
+Pusher automatically closes connections when a user navigates to another web page or closes their
+web browser.
+If you need to close a client connection manually,
+refer to [this page](https://pusher.com/docs/client_api_guide/client_connect#disconnecting).
```
Comment (ryanmcginnis) [inline tutorials/appengine-pusher/index.md]: Change to "refer to Pusher's [Disconnecting from Pusher](https://pusher.com/docs/client_api_guide/client_connect#disconnecting) documentation"

**Pull request merged**